### ISSUE TITLE
Selenium Grid not working with WebView2 application

### ISSUE THREAD

**Issue by pramodkeeliputti12345** [SeleniumHQ/selenium#9714]
## 🐛 Bug Report

Selenium Grid not working with WebView2 application. 

## To Reproduce

Detailed steps to reproduce the behavior:
Run the below c# code 
```csharp
            EdgeOptions edgeOptions = new EdgeOptions();
            edgeOptions.UseChromium = true;
            edgeOptions.UseWebView = true;
            edgeOptions.BinaryLocation = @"C:\MyWebView2App.exe";
            IWebDriver webDriver = new RemoteWebDriver(new Uri("http://127.0.0.1:4444"),edgeOptions);
```
Below exception occurs:
```
OpenQA.Selenium.WebDriverException
  HResult=0x80131500
  Message=The HTTP request to the remote WebDriver server for URL http://127.0.0.1:4444/session timed out after 60 seconds.
  Source=WebDriver
  StackTrace:
   at OpenQA.Selenium.Remote.HttpCommandExecutor.MakeHttpRequest(HttpRequestInfo requestInfo)
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.StartSession(ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(ICommandExecutor commandExecutor, ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(Uri remoteAddress, ICapabilities desiredCapabilities, TimeSpan commandTimeout)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(Uri remoteAddress, ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(Uri remoteAddress, DriverOptions options)
   at WebViewTesting.Program.Main(String[] args) in 

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
WebException: The operation has timed out.
```
## Expected behavior
webDriver  is initialized and the application is launched 
## Test script or set of commands reproducing this issue

The sample test project (WebViewTesting) and sample app (SampleApp) can be found in below git repo.
https://github.com/pramodkeeliputti12345/pramodkeeliputti12345

1. Start selenium grid 4 (selenium-server-4.0.0-beta-4.jar) in standalone mode.
2. Load project WebViewTesting in visual studio 2019.
3. Copy 'SampleApp' folder under c drive.
4. Run the WebViewTesting application. 

Exception occurse @ line number 16. (IWebDriver webDriver = new RemoteWebDriver(new Uri("http://127.0.0.1:4444"),edgeOptions);)

What i am looking is to achieve the below functionality with Selenium Grid.
https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/webdriver

This works for EdgeDriver but not working for RemoteWebDriver.

## Environment

OS: <Windows 10
Browser: Webview2
Browser version: 91.0.864.37
Browser Driver version: 91.0.864.37
Language Bindings version: C# 
Selenium Grid version (if applicable): selenium-server-4.0.0-beta-4.jar

**Comment by diemol** [comment]
Can you please provide a [concise reproducible test case](http://sscce.org/), that includes the page required to execute the scenario (a sample page or a public site). Ideally a GitHub repository with all details.

**Comment by pramodkeeliputti12345** [comment]
The sample test project (WebViewTesting) and sample app (SampleApp) can be found in below git repo.
https://github.com/pramodkeeliputti12345/pramodkeeliputti12345

**Comment by diemol** [comment]
How do we reproduce this? There are no instructions in the GitHub repository.

**Comment by pramodkeeliputti12345** [comment]
i have added Instructions in the repo.

**Comment by pramodkeeliputti12345** [comment]
What i am looking is to achieve the below functionality with Selenium Grid.
https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/webdriver

This works for EdgeDriver but not working for RemoteWebDriver.

**Comment by diemol** [comment]
Is the binary also where the Grid is running? Is everything running on the same machine?

**Comment by pramodkeeliputti12345** [comment]
Yes. Everything running on the same machine.

**Comment by pramodkeeliputti12345** [comment]
@diemol : i can see the issue still has tag R-awaiting answer. Do you need any info regarding the issue?

**Comment by diemol** [comment]
Removed the label, however we have not had time to check this.

**Comment by diemol** [comment]
I had a look at this and the main issue is that `UseWebView` changes the browser name to `webview2`, that is why Grid won't match this request at all.

The payload body looks like this:
`{"browserName":"webview2","ms:edgeOptions":{"binary":"C:\\SampleApp\\WebView2WpfBrowser.exe"}}`

I made a couple of [changes](https://github.com/SeleniumHQ/selenium/commit/2d36ea2cd98af9ae33716e2591edd4e6c9cf126c) in order to let Grid match this case, and together a [custom driver config](https://www.selenium.dev/documentation/grid/configuration/toml_options/#configuring-and-customising-drivers) moved things forward.

This is the custom driver config, `webdriver-executable` needs to be the path where your MSEdgeDriver is.
```toml
[node]
detect-drivers = false
[[node.driver-configuration]]
display-name = "Edge WebView2"
stereotype = '{"browserName": "webview2"}'
webdriver-executable = 'C:\tools\WebDriverServers\msedgedriver.exe'
``` 

However, it did not work because I get `unknown error: DevToolsActivePort file doesn't exist`.

If I try the original code from https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/webdriver, and run everything locally:
```csharp
namespace WebViewTesting
{
    class Program
    {
        static void Main(string[] args)
        {
            EdgeOptions edgeOptions = new EdgeOptions();
            edgeOptions.UseWebView = true;
            edgeOptions.BinaryLocation = @"C:\SampleApp\WebView2WpfBrowser.exe";
            EdgeDriver driver = new EdgeDriver(edgeOptions);
            //Navigate the WebView2API Sample from bing.com to microsoft.com
            driver.Url = @"https://www.microsoft.com";
            //Exit Microsoft Edge Driver
            driver.Quit();            
        }
    }
}
```

I get the same error `unknown error: DevToolsActivePort file doesn't exist`.

If I try to create the session by posting the payload directly to MSEdgeDriver:
```json
{
	"capabilities": {
		"alwaysMatch": {
			"browserName": "webview2",
			"ms:edgeOptions": {"binary": "C:\\SampleApp\\WebView2WpfBrowser.exe"}
		}
	}
}
```

I get this response:
```json
{
	"value": {
		"error": "unknown error",
		"message": "unknown error: DevToolsActivePort file doesn't exist",
		"stacktrace": "Backtrace:\n\tMicrosoft::Applications::Events::EventProperties::unpack [0x00007FF7479E33C2+22610]\n\tMicrosoft::Applications::Events::time_ticks_t::time_ticks_t [0x00007FF747921C92+503778]\n\tOrdinal0 [0x00007FF747420FA5+659365]\n\tOrdinal0 [0x00007FF74744B0F4+831732]\n\tOrdinal0 [0x00007FF747446CAD+814253]\n\tOrdinal0 [0x00007FF747442842+796738]\n\tOrdinal0 [0x00007FF74747C6C6+1033926]\n\tOrdinal0 [0x00007FF747476DF3+1011187]\n\tOrdinal0 [0x00007FF74744DE97+843415]\n\tOrdinal0 [0x00007FF74744F368+848744]\n\tMicrosoft::Applications::Events::EventProperty::to_string [0x00007FF7475E8BC8+848072]\n\tMicrosoft::Applications::Events::EventProperty::EventProperty [0x00007FF7476DAB9D+39117]\n\tMicrosoft::Applications::Events::EventProperty::EventProperty [0x00007FF7476DDE12+52034]\n\tMicrosoft::Applications::Events::EventProperty::to_string [0x00007FF747520906+28166]\n\tMicrosoft::Applications::Events::time_ticks_t::time_ticks_t [0x00007FF747928921+531569]\n\tMicrosoft::Applications::Events::time_ticks_t::time_ticks_t [0x00007FF74792DE04+553300]\n\tMicrosoft::Applications::Events::time_ticks_t::time_ticks_t [0x00007FF74792DF95+553701]\n\tMicrosoft::Applications::Events::EventProperty::to_string [0x00007FF7475F6650+904016]\n\tBaseThreadInitThunk [0x00007FFA13397034+20]\n\tRtlUserThreadStart [0x00007FFA14722651+33]\n"
	}
}
```

Just for reference, this is the MSEdgeDriver I used, https://msedgedriver.azureedge.net/98.0.1108.56/edgedriver_win64.zip

To summarize, a couple of changes in Grid and the custom config are able to pass along the request to MSEdgeDriver properly. However, it does not work because it seems that an issue in MSEdgeDriver.

I am not really sure where issues to MSEdgeDriver are reported nowadays, but you can try to create one in their docs, probably they will redirect you to the right place.

I will close this issue as Grid has been changed to work but now looks like an issue in MSEdgeDriver.
